dotnet/winforms
# fix: `PictureBox.LoadAsync` throws `PlatformNotSupportedException`

Fixes #242
Fixes #1548


## Proposed changes

Async delegates have been deprecated in .NET Core (see https://github.com/dotnet/corefx/issues/5940) for reasons such as:
* Async delegates use deprecated IAsyncResult-based async pattern. This pattern is generally not supported throughout .NET Core base libraries.
* Async delegates depend on remoting (System.Runtime.Remoting) under the hood. Remoting is not supported in .NET Core.

For more detailed reasons and migration strategies please refer to https://devblogs.microsoft.com/dotnet/migrating-delegate-begininvoke-calls-for-net-core/


In case of WinForms, async pattern predates the Task-based Asynchronous Pattern and it does not return awaitables.
In order to minimise the change to the public API surface we continue to expose the existing API and just re-route the image loading routine to a `Task` runner under the covers.



## Customer Impact

- Customers can continue loading images using the existing API

## Regression? 

- Yes

## Risk

- Low, customers will be able to load image asynchronously again.







## Test methodology 

- manual via the integration app; the following use cases were considered
    - load an image that results in an error (e.g. https://publicdomainarchive.com/wp-content/uploads/2014/03/public-domain-images-free-stock-photos-high-quality-resolution-downloads-around-the-house-1.jpg - SSL error)
    - load a large image (e.g. https://initiate.alphacoders.com/download/wallpaper/685514/images5/jpg/228368184713846/97229) and observe a progress reported, as well as a UI responsiveness
    - cancel during a large image loading process and observe events correctly fired, no further progress reported, GC collecting memory




###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/1696)
